File: leetcode/205. Isomorphic Strings/s3.cpp
// OJ: https://leetcode.com/problems/isomorphic-strings/
// Author: github.com/punkfulw
// Ref: https://leetcode.com/problems/isomorphic-strings/discuss/57796/My-6-lines-solution
// Time: O(N)
// Space: O(1)
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        int mp1[128] = {0}, mp2[128] = {0};
        for (int i = 0; i < s.size(); i++){
            if (mp1[s[i]] != mp2[t[i]])
                return false;
            mp1[s[i]] = i + 1;
            mp2[t[i]] = i + 1;
        }
        return true;
    }
};

File: leetcode/59. Spiral Matrix II/s3.py
# OJ: https://leetcode.com/problems/spiral-matrix-ii/
# Author: github.com/punkfulw
# Ref: https://leetcode.com/problems/spiral-matrix-ii/discuss/22282/4-9-lines-Python-solutions
# Time: O(N^2)
# Space: O(1)
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        ans = [[0] * n for _ in range(n)]
        i, j, di, dj = 0, 0, 0, 1
        
        for k in range(n * n):
            ans[i][j] = k + 1
            if ans[(i + di) % n][(j + dj) % n] != 0:
                di, dj = dj, -di
            i += di
            j += dj
        return ans
